서버의 JSON 사진 목록을 받아 사진보관함에 다운로드하는 기능 구현

idevjinie opened the pull request:
> 새로 바꾼 점 (step 2 리팩토링)
1. PhotoDataSource 객체에서 사진첩의 변화가 생겼을때 NotifactionCenter.post 을 통해 PhotosViewController가 알도록 하고
PhotosViewController가 변화를 반영하도록 하였습니다. #12
2. ReusableView protocol를 만들고 채택하여 reuseIdentifier를 사용하도록 하였습니다. #11
> 고민하여 구현한 점 (step3)
1. DoodleDataSource객체가 생성될때 json 데이터를 받아와서 디코딩하도록 처리했고, codable 객체가 다 만들어지면 NotificationCenter.post을 이용해 변화를 알려 DoodleViewController가 컬렉션 뷰를 로드하도록 하였습니다. 
2. Codable 객체인 DoodleImageInfo의 속성 date를 처음엔 String으로 두었지만, JSONDecoder를 통해 Date객체로 만들수 있다는 걸 알게 되어 Date 객체로 만들었습니다. #15, #14
3. ImageCollectionCell이라는 protocol를 만들어서 채택하면 cellSize라는 정적 변수를 사용하도록 하였습니다. 
4. Notification.Name객체들을 한 곳에 모으지 않고 사용하는 클래스의 정적 상수로 두어 사용하게끔 하였습니다. 
5. 이미지 다운로드 할때 쓰는 비동기적으로 코드를 구현해야 하는 경우에서 가독성을 위해 if 문을 사용하지 않고 guard문을 사용하였습니다.
6. 스토리보드가 아닌 코드로 구현하다 보니 흐름 파악이 어려웠습니다. 컬렉션뷰의 셀에 커스텀 셀 클래스를 연결하는 부분을 놓쳤었는데 register 메서드로 DoodleCell과 ReuseIdentifier를 지정하였습니다. #10 
7. 특정 셀을 롱클릭할 경우 해당 셀이 아닌 다른 셀에 MenuItem이 표시되었는데, 타겟 범위를 해당 셀의 contentView의 frame으로 지정하여 해결하였습니다. #10

ehgud0670 committed:
Style: DataSource 역할을 하는 객체가 모델 객체라 폴더 Models를 만들고 이동시킴

ehgud0670 committed:
Feat: 네비게이션 바에 타이틀 Photos 추가함

ehgud0670 committed:
Feat: Codable을 채택한 모델 객체 DoodleImage 를 만든다.
모델 객체 DoodleImage 는 다음과 같은 속성을 갖는다.

let order: Int
let imageURL: URL
let date: Date

ehgud0670 committed:
Feat: json 데이터의 키와 싱크를 맞추도록 프로토콜 CodingKey를
      채택하는 enum CodingKeys 생성함

ehgud0670 committed:
Feat: date 속성을 Date 객체를 확장해서 yyyyMMdd 포맷에 맞게 처리할 수 있도록
      기능 구현함.

ehgud0670 committed:
Add .gitattributes 추가

idevjinie committed:
Style: 기존 ViewController 파일 및 클래스명을 PhotosViewController로 변경

idevjinie committed:
Feat: DoodleViewController 생성

idevjinie committed:
Feat: NavigationBar 좌측에 + 버튼 추가

ehgud0670 committed:
Refactor : 더이상 상속하지 않을 뷰컨트롤러들 final 선언함

ehgud0670 committed:
Feat: doodle 데이터가 있는 주소를 상수로 만들어서 사용함

idevjinie committed:
Feat: NavigationBar의 + 버튼 클릭 시 DoodleViewController를 Modal로 표시

ehgud0670 committed:
Feat: doodle 이미지 관련 JSON 데이터를 가져와서 Doodle 배열 객체로
      주입(처리) 하도록 함

ehgud0670 committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Refactor: 모델 객체가 JSON 데이터와 싱크가 되도록 타입들 바꿈

ehgud0670 committed:
Refactor: 승인이 결정이 안 난 경우 승인을 요청하고 요청값을
          탈출 클로저를 이용해 받아내고 요청값이 true이면 viewDidload을
	  다시 하도록 한다.

          이유: 이전의 코드는 절대 승인이 결정 안난 경우 다시 승인을 요청했을때
	  요청값을 받지 못하는 구조였다. ( request하는 것은 해당 함수가 끝난뒤에 일어나는
	  비동기함수라서) 따라서 반영되도록 고쳤다.

ehgud0670 committed:
Refactor: json 데이터를 모델 객체가 잘 처리하도록 리팩토링함
	이유 : 이전의 코드는 비동기함수인 URLSession.shared.dataTask
	       를 처리할 수 없는 구조였다.
	      따라서 탈출클로저를 사용하여 처리할 수 있도록 하였고
	      또 함수의 성격에 맞게 함수도 분리하였다.

ehgud0670 committed:
Refactor: URLSession 관련 메소드는 네트워크 관련이므로
	  네트워크 클래스를 만들어서 옮김

ehgud0670 committed:
Deleted: 필요없는 dateFormatter 속성 삭제함

idevjinie committed:
Feat: DoodleViewController가 UICollectionViewController를 상속받도록 함

ehgud0670 committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Deleted: 안쓰는 메소드 삭제함

idevjinie committed:
Feat: NavigationBar 우측에 dismiss 기능을 하는 Close 버튼 추가

ehgud0670 committed:
Refactor: protocol를 채택하여 reuseIdentifier를 사용하도록 함

ehgud0670 committed:
Refactor: 델리게이트 패턴 대신 옵저버 패턴 이용
	DataSource 역할을 하는 모델 객체는 하나의 객체만 참조하기 보단
	여러개의 객체를 참조할 가능성이 크다.
	따라서 일대일 관계인 델리게이트 패턴 대신
	일대다 관계를 만들수 있는 옵저버패턴으로 적용하였다.

	* 사진을 추가하여 제대로 동작하는지 확인하였습니다.

ehgud0670 committed:
Refactor: 클래스 이름을 DataDecoder로 정정 및 적용

ehgud0670 committed:
Refactor: DoodleImageManager 객체 만들어서 DoodleImages를 관리하도록 함

ehgud0670 committed:
Style: 파일 분리함
	이유 : DoodleCell에도 적용될 것 같아 파일 분리한다.

ehgud0670 committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

idevjinie committed:
Deleted: 테스트 용도인 테스트 프로젝트 삭제함

idevjinie committed:
Feat: DoodleViewController에 UIImageView를 가진 커스텀 셀 클래스 연결

ehgud0670 committed:
Refactor: DoodleImage 객체의 속성의 타입을 더 적절한 타입으로 수정함
 => date: String -> Date

Refactor: JSONDecoder로 DoodleImage 객체의 속성에 맞게 JSON 데이터가 처리되도록 함

idevjinie committed:
Feat: DoodleDataSource 생성

idevjinie committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Refactor: DateStrategy 가 옵션이 되도록 옵셔널 타입으로 변경함

ehgud0670 committed:
Feat: DoodleVC 가 뷰 로드할때 json데이터를 모델 객체에 매핑하도록 코드 구현함

ehgud0670 committed:
Feat: 일단은 cell에 이미지 반영되도록 구현함
	그런데 안되는 게 문제임

ehgud0670 committed:
Refactor: 이미지뷰 오토레이아웃 코드로 적용하도록 기본 설정함

ehgud0670 committed:
Refactor: extension이므로 extensions 폴더로 파일 이동 시킴

ehgud0670 committed:
Style: 클래스 이름을 더 적절하게 변경함

idevjinie committed:
Refactor: DoodleImage 구조체명을 DoodleImageInfo로 변경

ehgud0670 committed:
Refactor: 이미지를 잘 갖고오도록 리팩토링 함

ehgud0670 committed:
Refactor: 다운로드한 이미지 잘 반영되도록 코드 구현

idevjinie committed:
Feat: DoodleCell의 doodleImageView 제약 설정"

idevjinie committed:
Refactor: doodleImageView 프로퍼티의 lazy 제거

idevjinie committed:
Test

ehgud0670 committed:
Refactor: 다운로드한 이미지 셀에 제대로 반영되도록 리팩토링함

idevjinie committed:
Deleted: 불필요한 메서드 정리

idevjinie committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Refactor: 다운받은 이미지 개수가 40개이거나 다 다운받은 경우에만
	 reloadData() 을 하도록 기능 구현함

ehgud0670 committed:
Refactor: 가독성을 높이기 위해 if문 삭제하고 guard문 적용함

idevjinie committed:
Feat: UILongPressGestureRecognizer 생성

idevjinie committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Refactor: 따로 Extension 파일들을 모으지말고
사용하는 곳에 속성을 추가하거나 extension을 이동시킴

ehgud0670 committed:
Refactor: 콜렉션 뷰에서 Cell을 설정할때 이미지가 다운로드 되어 반영되도록 기능 구현함

ehgud0670 committed:
Refactor: ImageCollectionCell 을 만들고 각 Cell 클래스에 채택하게 하여
	  정적 변수인 cellSize를 강제적으로 갖도록 함

ehgud0670 committed:
Refactor: 개행 정리하고 가독성을 위해 함수 분리함

ehgud0670 committed:
Refactor: 가독성을 위해 함수를 분리하고 extension을 사용함

ehgud0670 committed:
Refactor: DoodleDataSource 객체가 생성될때 json데이터를 받아오도록 기능 수정하였고
	  DoodleDataSource 를 분리해서 반드시 있어야 하는 collectionView 메소드와
	  내가 만든 메소드를 분리함

ehgud0670 committed:
Refactor: 디스패치큐 main 비동기문을 이동시킴
	이유: 콜렉션 뷰를 리로드하는 것은 좀 더 제네릭해야 하기 때문에
	디스패치큐 main 비동기문을 이동시켰다.

idevjinie committed:
Feat: DoodleViewController에서 특정 셀을 롱클릭하면 Save 액션을 하는 UIMenuItem 표시

ehgud0670 committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

idevjinie committed:
Feat: Save 액션 선택 시 해당 이미지를 사진보관함에 저장

idevjinie committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Refactor: cell 사이즈의 width를 110으로 바꿈

idevjinie committed:
Refactor: DoodleCell의 크기 변경

ehgud0670 committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 committed:
Refacotor: 클로저로 속성 설정하는 것이 묶여지도록 리팩토링함

godrm commented on PhotosApp/PhotosApp/Assets.xcassets/github.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+    },
+    {
+      "idiom" : "universal",
+      "scale" : "3x"
```
이미지 에셋을 사용할 경우는 고배율을 사용하세요. 이제 1배율을 사용하는 디바이스는 없습니다.

idevjinie replied:
디바이스별 디스플레이와 이미지 에셋의 관계에 대해 학습하였습니다. 해당 테스트용 이미지는 삭제하였습니다.

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,110 @@
+
+enum URLInfo {
+    
+    static let addressAboutDoodleDatas = "https://public.codesquad.kr/jk/doodle.json"
```
외부에서 필요한게 아니라면 클래스 내부에 넣어도 좋습니다. 참고하세요.

ehgud0670 replied:
넵 반영하였습니다. 감사합니다.

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,110 @@
+            else {
+                return
+                
+        }
```
else return 한 줄 뿐이라면 굳이 내려쓰지 않아도 되지 않을까요 (이건 그냥 제 개인 취향일 수도 있지만, 다른 코드가 없는데 굳이 더 길게 할 필요는 없으니까요)

ehgud0670 replied:
아 넵, 제가 생각해도 return 한줄만 있는 경우에는 안 내려쓰는게 나을 것 같아요. 반영하였습니다. 감사합니다.

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -0,0 +1,110 @@
+    private func setObservers() {
+        NotificationCenter.default.addObserver(forName:
+            DoodleDataSource
+                .notifiactionDoodleImageInfosDidChange,
```
여러 줄로 내려쓰는 기준은 뭔가요? 상수까지 내려써야 하는 이유는 모르겠어요.

ehgud0670 replied:
아 99자를 넘어가서 내려썼습니다. 근데 생각해보니 더 가독성을 해치는 것 같습니다. 
```swift
        NotificationCenter.default.addObserver(forName: DoodleDataSource.notifiactionDoodleImageInfosDidChange,
                                               object: doodleDataSource,
                                               queue: nil) { [weak self] _ in
                                                DispatchQueue.main.async {
                                                    self?.reloadCollectionView()
                                                }
        }
```
이런식으로 바꿨습니다. 감사합니다.

godrm commented on PhotosApp/PhotosApp/Models/DoodleDataSource.swift:
```diff
@@ -0,0 +1,71 @@
+            DoodleCell.reuseIdentifier, for: indexPath) as! DoodleCell
+        let itemCount = indexPath.item
+        guard itemCount < doodleImages.count
+            else {
```
guard else 를 이렇게 길게 쓸거면 if else 를 쓰거나 반대로 하는게 좋을 것 같습니다. 
guard else는 else 에서 exit 을 하기 위한 방식입니다.

ehgud0670 replied:
넵넵 guard else에서 else문은 exit 하는 경우에만 코드를 작성하도록 하겠습니다. 
해당 코드는 if else문을 작성하였습니다. 감사합니다.

godrm commented on PhotosApp/PhotosApp/Models/DoodleDataSource.swift:
```diff
@@ -0,0 +1,71 @@
+                }
+                return doodleCell
+        }
+        doodleCell.setPhoto(image: doodleImages[itemCount])
```
- 다운로드 처리하는 로직이 DataSource에 있기 보다는 별도로 분리하는 게 좋겠습니다 .
- 모든 이미지를 배열로 갖고 있는 것은 좋은 선택이 아닙니다. 배열이 커지면 앱이 죽을 수도 있으니까요. 
그래서 셀과 매칭하는 데이터 구조가 별도로 있는 게 좋습니다.

ehgud0670 replied:
넵, NSCache를 적용해보도록 하겠습니다. 가능하면 디스크에 데이터를 저장하도록 하겠습니다.

godrm replied:
NSCache가 다운로드를 대신해주거나 관리해주는 건 아닙니다. 
앱에서 파일을 다운로드해서 어디에 저장하는지 공부해보세요.

ehgud0670 replied:
아 넵넵 👍

godrm commented on PhotosApp/PhotosApp/Models/PhotosDataSource.swift:
```diff
@@ -0,0 +1,112 @@
+    
+    private func sharePhotoLibraryChanges() {
+        PHPhotoLibrary.shared().register(self)
+    }
```
여기는 이미 기존에 구현한 거 아닌가요? 왜 이전꺼와 비교가 안되죠?

ehgud0670 replied:
아 기존에 구현했던건데 폴더 Models에 옮겨나서 비교가 안됐던 것 같습니당

godrm commented on PhotosApp/PhotosApp/Network/Network.swift:
```diff
@@ -0,0 +1,33 @@
+                else {
+                return
+            }
+            completion(data)
```
한 가지 팁이 있다면, 에러 났을 때 예외처리에 대한 흐름 제어가 꼭 있어야 합니다. 
print()를 쓰지 말고 디버거를 쓰세요. 
completion 핸들러 외에 error 핸들러가 있는 경우도 있습니다.

ehgud0670 replied:
아 에러났을때 디버거를 사용하면 print()을 안써도 되는 군요. 알겠습니다. print문은 지우도록 하고 에러 났을때 예외 처리를
해보도록 하겠습니다.

godrm commented on PhotosApp/PhotosApp/Utils/DataDecoder.swift:
```diff
@@ -0,0 +1,38 @@
+                    let T = try jsonDecoder.decode(T.self, from: data)
+                    completion(T)
+                } catch let error {
+                    print(error.localizedDescription)
```
예외 catch 흐름 제어를 꼭 추가하세요.

godrm commented on PhotosApp/PhotosApp/Views/DoodleCell.swift:
```diff
@@ -0,0 +1,68 @@
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        setupDoodleImageView()
+    }
```
불문율이지만 init()이 가장 먼저 불리니까 다른 메소드들 보다는 위에 선언하는 게 어떨까요?

idevjinie replied:
혹시 다른 메소드들이라는 것이 무엇을 지칭하는지 설명해주실 수 있나요?

godrm replied:
여기 init보다 위에 선언한 func 메소드들이요
파일을 위에서부터 읽는 순서를 호출되는 순서와 비슷하게 하는게 좋습니다

ehgud0670 replied:
func 메소드 없지 않나요 ㅜ 
혹시 연산 프로퍼티 `canBecomeFirstResponder`나 클로저로 생성시킬 `doodleImageView` 프로퍼티를 말씀하시는 건가요?

godrm replied:
어라 이 클래스가 아니라 다른 클래스에 대한 건가 
(여기는 computed property 뿐이네요) 
모바일에서 위에 코드가 안 보이는데 다른 클래스도 확인해보세요

ehgud0670 replied:
아 넵, 알겠습니다

ehgud0670 replied:
혹시 DoodleDataSource 클래스의 아래 코드 말씀하시는건가요??
```swift
    private var doodleImageInfos = [DoodleImageInfo]() {
        didSet {
            NotificationCenter.default.post(name: DoodleDataSource.notifiactionDoodleImageInfosDidChange,
                                            object: self)
        }
    }
```

ehgud0670 replied:
doodleImageInfos 프로퍼티를 사용하는 메소드 아래에 위치해둠으로써 문제를 해결하였습니다.

godrm commented on PhotosApp/PhotosApp/Views/DoodleCell.swift:
```diff
@@ -0,0 +1,68 @@
+    }
+    
+    func photo() -> UIImage? {
+        return doodleImageView.image
```
아.. 이건 좀 어색합니다. 
저장할 때 뷰에서 이미지를 가져가지 말고, 해당하는 인덱스로 모델에서 참조해서 처리하는 게 좋습니다. 
그러면 여기 photo() 메소드는 필요없겠죠.

idevjinie replied:
모델에서 해당 인덱스를 사용하여 처리하도록 변경하였습니다. 모델과 뷰의 역할에 유의하겠습니다.

godrm reviewed (changes requested):
다음 피드백을 확인해보세요.

idevjinie committed:
Deleted: 테스트용 이미지 삭제

ehgud0670 committed:
Refactor: 해당 url 정보가 클래스 내부에서만 필요하므로 클래스 내부 속성으로 옮김

ehgud0670 committed:
Refactor: guard else 문에서 return 문만 있다면 개행을 안하도록 코드 리팩토링함

idevjinie committed:
Refactor: 이미지 저장 시 해당 인덱스를 사용하여 처리하도록 개선

ehgud0670 committed:
Style: 개행 정리함

ehgud0670 committed:
Refactor: guard문 지우고 if문 사용 및 가독성을 위해 메소드 분리함

	* guard문 지우고 if문을 쓴 이유:
	  guard문의 else문을 exit하는 코드를 작성해야 하는 곳이지
	  로직을 작성해야 하는 곳이 아니다.
	  따라서 해당 코드의 else문에 exit하지 않고 로직을 작성해놔서
	  guard문 대신 if문을 사용하였다.

ehgud0670 committed:
Refactor: 이 클래스들은 상속 당하지 않을 것 같아서 성능 향상을 위해 final로 명시함

ehgud0670 committed:
Style: completionHandler가 처리한다는 의미에서 더 적절하여
	매개변수 명 변경함

ehgud0670 committed:
Chore: doodleImageInfos 프로퍼티의 위치를 옮김
        이유: didSet에 메소드를 호출하여 속성이 매우 길어졌는데
        init보다 위에 있어서 가독성을 해친다.
        따라서 해당 프로퍼티를 사용하는 메소드 아래에 위치해 둠으로써
	가독성을 높였다.

ehgud0670 committed:
Refactor: Self를 붙일 수 있는 정적변수는 Self로 수정함

ehgud0670 committed:
Merge branch 'feature-GCDScheduling' of https://github.com/codesquad-memeber-2020/photos-4 into feature-GCDScheduling

ehgud0670 commented:
예외 처리와 디스크에 저장하는 것을 제외하고 피드백 모두 반영하였습니다. 예외 처리와 디스크에 저장하는 것은 내일 해볼 예정입니다.

ehgud0670 committed:
[style] Network보단 NetworkManager가 역할을 맡고 있는 것 같아 클래스명 변경함

ehgud0670 committed:
[refactor] didSet 삭제하고 로드가 완료된 경우에 노티를 보내도록 수정함
	  이유: 해당 속성은 로드 할때 한 번만 바뀌기 때문에 didSet은 한번만 호출될 것이다.
	  따라서 로드 할때만 노티를 한번 보내면 되지 굳이 disSet을 사용할 필요가 없어서
          로드할때 노티를 보내도록 코드를 재수정하였다.

ehgud0670 committed:
[refactor] 로드한 이미지가 원래의 index로 셀이 존재하는 경우에만 적용되도록 기능 수정함
	   이유: 셀은 기본적으로 재사용되는 셀이라서 로드되는 중간에 스크롤하면
	   셸이 재사용되어 밑에 나타날 수 있는데 이때 로드한 이미지를 보여주는 것은 버그이다.
	   따라서 해당 셀의 index를 기준으로 cell을 찾아 리턴값이 nil이 아니고 cell이 존재한다면
	   그때 입혀준다.

ehgud0670 committed:
Refactor: cellForItem은 효율적이지 못해서 지우고 prepareReuse를 오버라이딩하였다.
또 setPhoto가 main.async에서 동작하도록 기능 수정하였다.

ehgud0670 committed:
[iOS] Feat: 이미지 캐싱하는 기능 구현함

ehgud0670 committed:
Chore: 개행 정리함

ehgud0670 committed:
[#featrue-GCDScheduling] refactor: 탈출클로저이므로 self를 weak하게 재선언함

ehgud0670 committed:
feat: deinit될때 observer를 제거하는 기능 구현

godrm commented on PhotosApp/PhotosApp/DoodleViewController.swift:
```diff
@@ -24,12 +25,40 @@ final class DoodleViewController: UICollectionViewController {
+            name: DoodleDataSource.notifiactionDoodleImageInfosDidChange,
+            object: doodleDataSource
+        )
+    }
```
여기 ViewController에서는 token을 지정하는 곳이 없는데요?

ehgud0670 replied:
아 실수했습니다. 다시 반영하도록 하겠습니다.

ehgud0670 replied:
반영했습니다.

godrm commented on PhotosApp/PhotosApp/PhotosViewController.swift:
```diff
@@ -39,19 +39,33 @@ final class PhotosViewController: UIViewController {
+    }
+    
+    private func setObservers() {
+        token = NotificationCenter.default.addObserver(forName: PhotosDataSource.notificationPhotoLibraryDidChange,
```
그리고보니 token 이라는 변수 이름도 어색하네요. observer 역할 아닌가요?

ehgud0670 replied:
네, 맞습니다. https://developer.apple.com/documentation/foundation/notificationcenter/1411723-addobserver 의 예제보고 아무생각 없이 token이라고 했네요 😅 수정하도록 하겠습니다. 👍

ehgud0670 replied:
반영했습니다.

ehgud0670 committed:
refactor:  observer 등록 및 token 이 observer 역할을 하므로 observer로 정정

godrm commented:
수고하셨습니다.

ehgud0670 commented:
사실 임시로 push 한거라 pr이 다시 살아날줄 몰랐습니다. ㅋㅋㅋ 
기왕 열려진거 지금 코드는 
1) Image Data를 프로퍼티로 가지고 있고, 
2) 이미지도 업데이트가 안되는데

메모리 낭비를 줄이기 위해 Data 프로퍼티를 삭제하고, 이미지도 업데이트될 수 있게끔만 구현해보도록 하겠습니다~!

godrm commented:
PR 안 닫힌 상태에서 그냥 Push 해서 그런가보네요. 이거 닫고 개선해서 다시 보내세요. 
그래야 구분이 될 것 같네요.

ehgud0670 commented:
아 넵넵
